### DISCUSSION TITLE
How can I force a provider to rebuild even if nothing is watching it?

### DISCUSSION THREAD

**Discussion by miklcct** [rrousselGit/riverpod#3772] [Q&A]
Let's say I have two providers:
FooProvider - which gets a stream
BarProvider - which watches FooProvider and produces a value based on the element watched

Both providers are always alive.

However, if nothing listens to BarProvider, BarProvider.build() isn't run even if FooProvider gets a new value. Why is that? I am using BarProvider which does some lengthy computation on top of FooProvider and I don't want a delay when BarProvider is read.

**Comment by AhmedLSayed9** [comment]


**Reply by miklcct** [reply]
I don't know because my BarProvider is very complicated, but once I put a listen in the main widget for BarProvider it works as I expect.

The exact provider is this, which attempts to do reverse geocoding on a stream of position:

```dart
@Riverpod(keepAlive: true)
class CurrentPlace extends _$CurrentPlace {
  (Place, DateTime)? lastPlace;
  final _throttler = Throttler();

  static const namePlaceholder = 'Current location';

  @override
  Place? build() {
    final reverseGeocode = ref.watch(reverseGeocodeProvider);
    final position = ref.watch(currentPositionProvider);
    final lastPlace = this.lastPlace;
    if (position != null &&
        (lastPlace == null ||
            (const Distance()).distance(lastPlace.$1.position, position) >=
                100 ||
            DateTime.now()
                .isAfter(lastPlace.$2.add(const Duration(seconds: 30))))) {
      _throttler.throttle(
          duration: const Duration(seconds: 5),
          onThrottle: () async {
            try {
              final place = await reverseGeocode(position);
              this.lastPlace = (place, DateTime.now());
              ref.invalidateSelf();
            } on Exception catch (e) {
              if (e is! geocoding.NoResultFoundException) {
                debugPrint(e.toString());
              }
              this.lastPlace = null;
            }
          });
    }
    return position?.pipe((position) => Place(
        name: lastPlace?.$1.name ?? namePlaceholder,
        city: lastPlace?.$1.city,
        position: position));
  }
}
```

I attached a debugger on the build method and it only ran once when the provider was read, but not afterwards even if positions were received.

**Reply by AhmedLSayed9** [reply]
@miklcct 
I guess you've never read `BarProvider` at your code?
You've to read it at least one time as providers are lazy loaded by default.

**Reply by miklcct** [reply]
I had read it once immediately in my root widget using a useEffect call. After I changed the read to listen it behaved as I expected.

**Comment by AhmedLSayed9** [comment]
Well, that's expected.
If a keep-alive provider is disposed (a rebuild) for any reason i.e ref.watch/ref.invalidate, and there're no other listeners to it, it'll not auto build again and instead will back to the original state.
